test/index.js

```diff
@@ -460,4 +470,81 @@ describe('KeyringController', function () {
       );
     });
   });
+
+  describe('submitEncryptionKey', function () {
+    it('sets encryption key data upon submitPassword', async function () {
+      keyringController.cacheEncryptionKey = true;
+      await keyringController.submitPassword(password);
+
+      expect(keyringController.password).toBe(password);
+      expect(keyringController.encryptionSalt).toBe('SALT');
+    });
+
+    it('unlocks the keyrings with valid information', async function () {
+      keyringController.cacheEncryptionKey = true;
+      const returnValue = await keyringController.encryptor.decryptWithKey();
+      const stub = sinon.stub(keyringController.encryptor, 'decryptWithKey');
+      stub.resolves(Promise.resolve(returnValue));
+
+      keyringController.store.updateState({ vault: MOCK_ENCRYPTION_DATA });
+      await keyringController.submitEncryptionKey(MOCK_ENCRYPTION_KEY);
+
+      expect(keyringController.encryptor.decryptWithKey.calledOnce).toBe(true);
+      expect(keyringController.keyrings).toHaveLength(1);
+    });
+
+    it('should not load keyrings when incorrect encryptionKey', async function () {
+      keyringController.cacheEncryptionKey = true;
+      keyringController.setLocked(true);
+
+      await expect(keyringController.submitEncryptionKey(`{}`)).rejects.toThrow(
+        `Failed to execute 'importKey' on 'SubtleCrypto': The provided value is not of type '(ArrayBuffer or ArrayBufferView or JsonWebKey)'.`,
+      );
+      expect(keyringController.password).toBeUndefined();
+      expect(keyringController.keyrings).toHaveLength(0);
+    });
+
+    it('persists keyrings when actions are performed', async function () {
+      keyringController.cacheEncryptionKey = true;
+      keyringController.store.updateState({ vault: MOCK_ENCRYPTION_DATA });
+      await keyringController.submitEncryptionKey(MOCK_ENCRYPTION_KEY);
+
+      const stub = sinon.stub(keyringController, 'persistAllKeyrings');
+      stub.resolves(Promise.resolve(true));
```

Comment:
If you just want to see how a function was called, you don't need to stub it out. Instead use `sinon.spy`, which lets you spy on the function without changing how it behaves.

```suggestion
      const spy = sinon.spy(keyringController, 'persistAllKeyrings');
```

Reply:
Though... generally it's not a good idea to be using internals in tests. They can serve as an unnecessary obstacle to refactoring, and generally they're farther away from testing real use cases. If we could update the assertions to avoid using internals altogether that would be idea.

In this case, the `memStore` can be inspected to see how many accounts there are - there's no need to use a spy. And the same goes for the addition of a new keyring and the removal of an account. It's all in the memStore.

Reply:
Good shout!